### PR TITLE
Warn if the first argument of an instance/ class method gets assigned, closes #977

### LINKED ISSUE #977: Add test for assignments to self/cls in class methods
See `find_self_assignments` [in test_code_quality.py](https://github.com/diofant/diofant/blob/b4a1914/sympy/utilities/tests/test_code_quality.py) as an example of implementation.  This will catch code like

``` python
class A:
    def foo(self, arg):
        self = arg
```

or

``` python
class A(object):
    @classmethod
    def foo(cls, arg):
        cls = arg
```

Previous attempt for this feature request was [this](https://github.com/PyCQA/pep8-naming/issues/17).  I hope now this issue is in the right place.

### PR DESCRIPTION
**Pull request by mar-chi-pan**
### Fixes / new features
- New warning if the first argument of an instance/ class method gets assigned

### CODE REVIEW

File: pylint/checkers/variables.py [hunk 1]
```diff
@@ -1325,6 +1334,24 @@ def visit_assign(self, node):
...
+        """Check that self/cls don't get assigned"""
+        scope = node.scope()
+        if not (isinstance(scope, astroid.scoped_nodes.FunctionDef) and
+                scope.is_method() and
+                "builtins.staticmethod" not in scope.decoratornames()):
+            return
+        self_cls_name_list = scope.argnames()[:1]
+        if not self_cls_name_list:
+            return
+        self_cls_name = self_cls_name_list.pop()
```
Comment (AWhetter) [inline pylint/checkers/variables.py]: Minor point: We could remove the need for slicing if we do something like

```python
argument_names = scope.argnames()
if not argument_names:
    return
self_cls_name = argument_names[0]
```

File: pylint/checkers/variables.py [hunk 2]
```diff
@@ -346,6 +346,10 @@ def _has_locals_call_after_node(stmt, scope):
               'Used when a variable is defined but might not be used. '
               'The possibility comes from the fact that locals() might be used, '
               'which could consume or not the said variable'),
+    'W0642': ('Invalid assignment to %s in method',
+              'self-cls-assignment',
+              'Invalid assignment to self or cls in instanct or class method '
```
Comment (AWhetter) [inline pylint/checkers/variables.py]: Typo: Should "instanct" be "instance"?

File: ChangeLog [hunk 3]
```diff
@@ -4,10 +4,14 @@ Pylint's ChangeLog
 What's New in Pylint 2.0?
 =========================
 
+    * Warn if the first argument of an instance/ class method gets assigned
+
+      CLose #977
```
Comment (AWhetter) [inline ChangeLog]: Typo: I'm being super picky, but the "L" should be lower case.

File: pylint/test/unittest_checker_variables.py [hunk 4]
```diff
@@ -100,6 +100,49 @@ def func():
         with self.assertAddsMessages(msg):
             self.checker.visit_global(node)
 
+    def test_self_assignment(self):
```
Comment (PCManticore) [inline pylint/test/unittest_checker_variables.py]: Hey @mar-chi-pan can you rewrite these as  functional tests instead?

Comment (PCManticore) [reply pylint/test/unittest_checker_variables.py]: Actually you already have functional tests so these could be deleted

**Pull request merged**